### PR TITLE
feat(route53-targets): aws-apigatewayv2 target

### PR DESCRIPTION
**Pull request by johanneswuerbach**
Add support for [aws-apigatewayv2](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-apigatewayv2-readme.html#custom-domain) `DomainName` as target.

Closes https://github.com/aws/aws-cdk/issues/8941

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-route53-targets/lib/api-gatewayv2-domain-name.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import * as apigv2 from '@aws-cdk/aws-apigatewayv2';
+import * as route53 from '@aws-cdk/aws-route53';
+
+/**
+ * Defines an API Gateway V2 domain name as the alias target.
+ */
+export class ApiGatewayv2Domain implements route53.IAliasRecordTarget {
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-route53-targets/lib/api-gatewayv2-domain-name.ts:7]: curious: how does the casing here translate for Python? You can run `yarn package` to generate the distributions for other languages.

would it be `api_gatewayv2_domain` ? wondering if it should be `ApiGatewayV2`

File: packages/@aws-cdk/aws-route53-targets/README.md [hunk 2]
```diff
@@ -16,6 +16,13 @@ This library contains Route53 Alias Record targets for:
     // or - route53.RecordTarget.fromAlias(new alias.ApiGatewayDomain(domainName)),
   });
   ```
+* API Gateway V2 custom domains
+  ```ts
+  new route53.ARecord(this, 'AliasRecord', {
+    zone,
+    target: route53.RecordTarget.fromAlias(new alias.ApiGatewayv2Domain(domainName)),
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-route53-targets/README.md:23]: can the default domain of `ApiGatewayv2` be used as a target? (asking if we also need a class like `ApiGateway` for that scenario

File: packages/@aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts [hunk 3]
```diff
@@ -5,7 +5,7 @@ import * as route53 from '@aws-cdk/aws-route53';
  * Defines an API Gateway domain name as the alias target.
  *
  * Use the `ApiGateway` class if you wish to map the alias to an REST API with a
- * domain name defined throug the `RestApiProps.domainName` prop.
+ * domain name defined through the `RestApiProps.domainName` prop.
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts:8]: appreciate the cleanup here (and elsewhere in this PR!).
in the future, let's save these for a different PR so they don't take focus away.

**Review by shivlaks — CHANGES_REQUESTED** [review]
thanks for getting this started @johanneswuerbach 

I had some questions and some small comments. 
I think we should also add an integration test for this target.

File: packages/@aws-cdk/aws-route53-targets/README.md (hunk 2 above)
Comment (johanneswuerbach) [reply packages/@aws-cdk/aws-route53-targets/README.md:23]: I also originally planned to provide something similar, but unlike the v1 version where you can just [provide a domain name string](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-apigateway-readme.html#custom-domains), v2 always requires domain name instances to be [created explicitly](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-apigatewayv2-readme.html#custom-domain).

This default domain name parameter is also then [setting the domain name of a default stage](https://github.com/aws/aws-cdk/blob/ddd29b485d5e07c0661ce135460b34d9f2b712e9/packages/%40aws-cdk/aws-apigatewayv2/lib/http/api.ts#L186-L190), which seemed fairly implicit as a target.

wdyt?

File: packages/@aws-cdk/aws-route53-targets/lib/api-gatewayv2-domain-name.ts (hunk 1 above)
Comment (johanneswuerbach) [reply packages/@aws-cdk/aws-route53-targets/lib/api-gatewayv2-domain-name.ts:7]: Is there a way to easily run this without installing `yarn`, `javac` and others? Like a docker image?

File: packages/@aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts (hunk 3 above)
Comment (johanneswuerbach) [reply packages/@aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts:8]: Sorry, happy to remove it just noticed the typo when copying the existing target.

File: packages/@aws-cdk/aws-route53-targets/lib/api-gatewayv2-domain-name.ts (hunk 1 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-route53-targets/lib/api-gatewayv2-domain-name.ts:7]: `yarn package`, but class names are fine - ignore this comment!

File: packages/@aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts (hunk 3 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-route53-targets/lib/api-gateway-domain-name.ts:8]: it's fine to leave in, just a note for future contributions! 🙂

File: packages/@aws-cdk/aws-route53-targets/README.md (hunk 2 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-route53-targets/README.md:23]: gotcha, thanks for clarifying, I think this is fine as it is in that case.

**Review by shivlaks — DISMISSED** [review]
@johanneswuerbach could you also add an integration test?

**Comment by johanneswuerbach** [comment]
Sure, could you point me to an example test?

**Comment by christophgysin** [comment]
@johanneswuerbach Here's an example integration test:
https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-route53-targets/test/integ.api-gateway-domain-name.ts

**Comment by johanneswuerbach** [comment]
@christophgysin thanks for the pointer.

@shivlaks integration test added.

File: packages/@aws-cdk/aws-route53-targets/test/integ.api-gatewayv2-domain-name.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import * as targets from '../lib';
+
+class TestStack extends Stack {
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    const domainName = 'example.com';
+    const certArn = 'arn:aws:acm:us-east-1:111111111111:certificate';
+
+    const certificate = acm.Certificate.fromCertificateArn(this, 'cert', certArn);
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-route53-targets/test/integ.api-gatewayv2-domain-name.ts:16]: integration tests should be deployable (i.e. running `npm integ *your test.js*` will deploy the stack and automatically write the `.expected` file)

should this test be creating a certificate and then passing it's ARN? I can't imagine it deploys with the hardcoded `certArn` value

**Review by shivlaks — CHANGES_REQUESTED** [review]
code looks good, I think we need to tweak the integ test a little bit before we can merge it in

**Comment by christophgysin** [comment]
@shivlaks I haven't tried it, but this is exactly how it is done in the existing test: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-route53-targets/test/integ.api-gateway-domain-name.ts#L15

**Comment by shivlaks** [comment]
> @shivlaks I haven't tried it, but this is exactly how it is done in the existing test: https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-route53-targets/test/integ.api-gateway-domain-name.ts#L15

if it's not deployable it's probably also a bug (or there for a reason). I'd love to find out whether we can do better before merging it in though. wdyt?

**Comment by christophgysin** [comment]
I'm neither the author of this nor the existing test. But I can imagine that the certificate ARN and domain name were changed to an existing cert/domain in the account for tedting.

Automating certificate creation is not recommended, because you quickly run into the account limit of 20 certificates per year.

**Comment by johanneswuerbach** [comment]
Anything I can do to move this forward?

**Comment by christophgysin** [comment]
@shivlaks Can you comment how to proceed here? Should we automate certificate creation despite the limits? Is there another way to test this? Or this the current approach (not ideal, but consistent with targets.ApiGateway) acceptable in this case?

**Comment by shivlaks** [comment]
@christophgysin @johanneswuerbach - discussed the outstanding integ test comment with @njlynch - and we agreed that it's probably better to drop it if it's not executable. The rationale is that it doesn't really buy us much protection / validation that cannot be covered through unit tests themselves. 

So: let's drop it and ship this PR!!

**Comment by johanneswuerbach** [comment]
@shivlaks thanks for the update, integration tests removed.

**Review by shivlaks — APPROVED** [review]

**Pull request merged**